designs/2024-multithread-linting/README.md

```diff
@@ -0,0 +1,882 @@
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+When the `concurrency` option is not used, nothing will change.
+When the `concurrency` option is used, some things will change:
+* Unserializable `ESLint` constructor options will throw an error. This does not affect CLI options.
+* Some errors thrown by `ESLint#lintFiles()` could be different (see [error management](#error-management)).
+* Tools that rely on `ESLint#getRulesMetaForResults()` may start behaving incorrectly, depending on the approach chosen in [`ESLint#getRulesMetaForResults()`](#eslintgetrulesmetaforresults).
+* typescript-eslint will still work, but if type-aware linting is used there will be considerable additional overhead because each thread needs to load the TypeScript project.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+There is probably no single alternative that would address all use cases considered in this document.
+However some of the problems tackled in this RFC can be solved in different ways.
+Some of these alternatives are described closer in the single sections of this RFC.
+
+For a tool, the alternative I looked into more closely is [Trunk Code Quality](https://trunk.io/code-quality), one of the tools used to build ESLint.
+Trunk is able to parallelize the lint process by splitting ESLint runs over multiple processes.
+However, it does so in a way that makes it difficult to collect the lint results in the end of the process.
+This ability is essential for many core functionalities to work, for example formatters.
+Also Trunk does not offer an API similar to ESLint to control the lint process programmatically.
+
+Another interesting tool is a fork of ESLint v8 with parallel linting support shared by [faultyserver](https://github.com/faultyserver)[^4].
+This tool only supports eslintrc configuration which makes it challenging to directly compare the results with the current single-thread implementation.
+
+## Open Questions
+
+<!--
+    This section is optional, but is suggested for a first draft.
+
+    What parts of this proposal are you unclear about? What do you
+    need to know before you can finalize this RFC?
+
+    List the questions that you'd like reviewers to focus on. When
+    you've received the answers and updated the design to reflect them, 
+    you can remove this section.
+-->
+
+* Are there any viable alternatives to [option modules](#option-modules)?
+* How to serialize CLI options? (see [Solution for CLI Options](#solution-for-cli-options) for a proposed solutions).
+* How to fix `ESLint#getRulesMetaForResults()` to support multithreading? (see proposed solutions in [`ESLint#getRulesMetaForResults()`](#eslintgetrulesmetaforresults)).
```

Comment:
It looks like these are addressed in the RFC so we can move out of open questions.

Reply:
Okay then, we can try the different approaches in the prototyping phase and decide what is working better.